hazelcast/src/main/java/com/hazelcast/jet/impl/JetClientInstanceImpl.java

```diff
@@ -149,96 +143,89 @@ public Job newJobProxy(long jobId, boolean isLightJob, @Nonnull Object jobDefini
+            boolean result = sendJobMetaData(jobUploadCall, submitJobParameters);
             if (result) {
                 logFine(getLogger(), "Submitted JobMetaData successfully for jarPath: %s", jarPath);
             }
             // Send job parts
-            sendJobMultipart(jarPath, sessionId, jarSize);
+            sendJobMultipart(jobUploadCall, jarPath);
 
         } catch (IOException | NoSuchAlgorithmException exception) {
             sneakyThrow(exception);
         }
     }
 
-    // Validate the parameters used by the client
-    protected void validateParameterObject(SubmitJobParameters parameterObject) {
-        // Check that parameter is not null, because it is used to access the file
-        if (Objects.isNull(parameterObject.getJarPath())) {
-            throw new JetException("jarPath can not be null");
-        }
+    // This method is public for testing purposes.
+    public JobUploadCall initializeJobUploadCall(Path jarPath)
+            throws IOException, NoSuchAlgorithmException {
 
-        // Check that parameter is not null, because it is used by the JetUploadJobMetaDataCodec
-        if (Objects.isNull(parameterObject.getJobParameters())) {
-            throw new JetException("jobParameters can not be null");
-        }
-    }
+        JobUploadCall jobUploadCall = new JobUploadCall();
+        jobUploadCall.initializeJobUploadCall(client, jarPath);
 
-    // This method is public for testing purposes.
-    public String getFileNameWithoutExtension(Path jarPath) {
-        String fileName = jarPath.getFileName().toString();
-        if (!fileName.endsWith(".jar")) {
-            throw new JetException("File name extension should be .jar");
-        }
-        fileName = fileName.substring(0, fileName.lastIndexOf('.'));
-        return fileName;
+        return jobUploadCall;
     }
 
-    private boolean sendJobMetaData(UUID sessionId, String fileNameWithoutExtension, String sha256Hex,
+    private boolean sendJobMetaData(JobUploadCall jobUploadCall,
                                     SubmitJobParameters submitJobParameters) {
-        ClientMessage jobMetaDataRequest = JetUploadJobMetaDataCodec.encodeRequest(sessionId,
-                fileNameWithoutExtension, sha256Hex,
+        ClientMessage jobMetaDataRequest = JetUploadJobMetaDataCodec.encodeRequest(
+                jobUploadCall.getSessionId(),
+                jobUploadCall.getFileNameWithoutExtension(),
+                jobUploadCall.getSha256HexOfJar(),
                 submitJobParameters.getSnapshotName(),
                 submitJobParameters.getJobName(),
                 submitJobParameters.getMainClass(),
                 submitJobParameters.getJobParameters());
 
-        return invokeRequestOnMasterAndDecodeResponse(jobMetaDataRequest, JetUploadJobMetaDataCodec::decodeResponse);
+        return invokeRequestAndDecodeResponseNoRetryOnRandom(jobUploadCall.getMemberUuid(), jobMetaDataRequest,
```

Comment:
If the client is unisocket and has no connection it means it has no connection at all. If unisocket has a single connection then it will use that one.
The new requirement for smart clients is that the upload should not always go to master. That's the reason of the change

Reply:
But we are getting the member uuid to invoke the request from the memberlist. In the case of a multi-member cluster, we might randomly select a member to which the unisocket client is not connected to. A similar thing applies for the smart client as well. We might select a random member that we don't yet connected (or lost the connection for a brief period of time). Maybe we can utilize the connection manager while selecting the connection to send this request to

Reply:
After a discussion over the private messages, we have concluded that there are no problems for the unisocket client, as we always invoke the connection we have. We concluded that the problem for the smart client cannot be avoided no matter what we do, so I am resolving this comment